oauth2client/transport/urllib3.py

```diff
@@ -0,0 +1,165 @@
+# Copyright 2016 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Transport implementation using urllib3"""
+
+from __future__ import absolute_import
+
+import logging
+
+import urllib3
+from urllib3 import request
+
+from oauth2client.transport import _helpers
+
+
+_LOGGER = logging.getLogger(__name__)
+_MAX_REFRESH_ATTEMPTS = 2
+
+
+def get_http_object(*args, **kwargs):
+    """Return a new HTTP object.
+
+    Args:
+        *args: tuple, The positional arguments to be passed when
+               contructing a new HTTP object.
+        **kwargs: dict, The keyword arguments to be passed when
+                  contructing a new HTTP object.
+
+    Returns:
+        urllib3.PoolManager: The new HTTP object.
+    """
+    return urllib3.PoolManager(*args, **kwargs)
+
+
+class AuthorizedHttp(request.RequestMethods):
+    """An authorized urllib3 HTTP class.
+
+    Implements :class:`urllib3.request.RequestMethods` and can be used just
+    like any other :class:`urllib3.PoolManager`.
+
+    Provides an implementation of :meth:`urlopen` that handles adding the
+    credentials to the request headers and refreshing credentials as needed.
+
+    Args:
+        http (urllib3.PoolManager): The underlying HTTP object to
+            use to make requests.
+        credentials (oauth2client.client.OAuth2Credentials): The credentials to
+            add to the request.
+        refresh_status_codes (Sequence): Which HTTP status code indicate that
+            credentials should be refreshed and the request should be retried.
+    """
+    def __init__(self, http, credentials, refresh_status_codes):
+        self.http = http
+        self.credentials = credentials
+        self.refresh_status_codes = refresh_status_codes
+
+    def urlopen(self, method, url, body=None, headers=None,
+                _credential_refresh_attempt=0, **kwargs):
+        """Implementation of urllib3's urlopen."""
+        if headers is None:
+            headers = self.headers
+
+        # Copy the request headers because this may be called recursively.
+        request_headers = _helpers.initialize_headers(headers).copy()
```

Comment:
Good point, the `.copy()` was not needed to fix the behavior that I was trying to avoid. I just needed to make sure to pass on the original headers instead of the modified copy. Fixed.